// @ton/sandbox/dist/ui/connection/websocket/ManagedWebSocketConnector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManagedWebSocketConnector = void 0;
const environment_1 = require("../../../utils/environment");
const constants_1 = require("./constants");
class ManagedWebSocketConnector {
    websocketAddress;
    ws;
    constructor({ wsPort = constants_1.DEFAULT_PORT, wsHost = constants_1.DEFAULT_HOST } = {}) {
        this.websocketAddress = (0, environment_1.getOptionalEnv)('SANDBOX_WEBSOCKET_ADDR') ?? `ws://${wsHost}:${wsPort}`;
    }
    async connect() {
        try {
            this.ws = await new Promise((resolve, reject) => {
                const ws = new WebSocket(this.websocketAddress);
                ws.addEventListener('open', () => resolve(ws), { once: true });
                ws.addEventListener('error', (err) => {
                    ws.close();
                    reject(err);
                }, { once: true });
            });
        }
        catch (err) {
            // eslint-disable-next-line no-console
            console.warn(constants_1.CONNECT_ERROR_MESSAGE, err);
        }
    }
    disconnect() {
        this.ws?.close();
        this.ws = undefined;
    }
    send(data) {
        if (this?.ws?.readyState === WebSocket.OPEN) {
            this.ws.send(data);
        }
    }
}
exports.ManagedWebSocketConnector = ManagedWebSocketConnector;
