// @ton/sandbox/dist/coverage/test/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeInstructions = executeInstructions;
const ton_assembly_1 = require("ton-assembly/");
const core_1 = require("@ton/core");
const __1 = require("../../");
async function executeInstructions(code, id = 0) {
    class TestContract {
        address;
        init;
        constructor(address, init) {
            this.address = address;
            this.init = init;
        }
        async send(provider, via, args, body) {
            await provider.internal(via, { ...args, body: body });
        }
        async getAny(provider, id) {
            const builder = new core_1.TupleBuilder();
            const res = (await provider.get(id, builder.build()));
            return [res.stack, res.vmLogs];
        }
    }
    const blockchain = await __1.Blockchain.create();
    blockchain.verbosity.print = false;
    blockchain.verbosity.vmLogs = "vm_logs_verbose";
    const treasure = await blockchain.treasury("treasure");
    const init = {
        code: ton_assembly_1.runtime.compileCell(code),
        data: new core_1.Cell(),
    };
    const address = (0, core_1.contractAddress)(0, init);
    const contract = new TestContract(address, init);
    const openContract = blockchain.openContract(contract);
    await openContract.send(treasure.getSender(), {
        value: (0, core_1.toNano)("10"),
    }, new core_1.Cell());
    return openContract.getAny(id);
}
