// ton-assembly/dist/runtime/fift-instr-constructors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fIF = exports.fTHROWIFNOT = exports.fTHROWIF = exports.fTHROW = exports.fPREPAREDICT = exports.fJMPDICT = exports.fCALLDICT = exports.fCALLXARGS = exports.fSDBEGINSQ = exports.fSDBEGINS = exports.fPLDUZ = exports.fPUSHINTX = exports.fXCHG = exports.fSTSLICECONST = exports.fPUSHCONT = exports.fPUSHSLICE = exports.fPOP = exports.fPUSH = exports.fPUSHINT = void 0;
const fPUSHINT = (arg0, loc) => ({
    $: "fPUSHINT",
    arg0,
    loc,
});
exports.fPUSHINT = fPUSHINT;
const fPUSH = (kind, arg0, loc) => ({
    $: "fPUSH",
    arg0,
    kind,
    loc,
});
exports.fPUSH = fPUSH;
const fPOP = (kind, arg0, loc) => ({
    $: "fPOP",
    arg0,
    kind,
    loc,
});
exports.fPOP = fPOP;
const fPUSHSLICE = (arg0, loc) => ({
    $: "fPUSHSLICE",
    arg0,
    loc,
});
exports.fPUSHSLICE = fPUSHSLICE;
const fPUSHCONT = (arg0, loc) => ({
    $: "fPUSHCONT",
    arg0,
    loc,
});
exports.fPUSHCONT = fPUSHCONT;
const fSTSLICECONST = (arg0, loc) => ({
    $: "fSTSLICECONST",
    arg0,
    loc,
});
exports.fSTSLICECONST = fSTSLICECONST;
const fXCHG = (arg0, arg1, loc) => ({
    $: "fXCHG",
    arg0,
    arg1,
    loc,
});
exports.fXCHG = fXCHG;
const fPUSHINTX = (arg0, loc) => ({
    $: "fPUSHINTX",
    arg0,
    loc,
});
exports.fPUSHINTX = fPUSHINTX;
const fPLDUZ = (arg0, loc) => ({
    $: "fPLDUZ",
    arg0,
    loc,
});
exports.fPLDUZ = fPLDUZ;
const fSDBEGINS = (arg0, loc) => ({
    $: "fSDBEGINS",
    arg0,
    loc,
});
exports.fSDBEGINS = fSDBEGINS;
const fSDBEGINSQ = (arg0, loc) => ({
    $: "fSDBEGINSQ",
    arg0,
    loc,
});
exports.fSDBEGINSQ = fSDBEGINSQ;
const fCALLXARGS = (arg0, arg1, loc) => ({
    $: "fCALLXARGS",
    arg0,
    arg1,
    loc,
});
exports.fCALLXARGS = fCALLXARGS;
const fCALLDICT = (arg0, loc) => ({
    $: "fCALLDICT",
    arg0,
    loc,
});
exports.fCALLDICT = fCALLDICT;
const fJMPDICT = (arg0, loc) => ({
    $: "fJMPDICT",
    arg0,
    loc,
});
exports.fJMPDICT = fJMPDICT;
const fPREPAREDICT = (arg0, loc) => ({
    $: "fPREPAREDICT",
    arg0,
    loc,
});
exports.fPREPAREDICT = fPREPAREDICT;
const fTHROW = (arg0, loc) => ({
    $: "fTHROW",
    arg0,
    loc,
});
exports.fTHROW = fTHROW;
const fTHROWIF = (arg0, loc) => ({
    $: "fTHROWIF",
    arg0,
    loc,
});
exports.fTHROWIF = fTHROWIF;
const fTHROWIFNOT = (arg0, loc) => ({
    $: "fTHROWIFNOT",
    arg0,
    loc,
});
exports.fTHROWIFNOT = fTHROWIFNOT;
const fIF = (kind, trueBranch, falseBranch, loc) => ({
    $: "fIF",
    kind,
    trueBranch,
    falseBranch,
    loc,
});
exports.fIF = fIF;
