// @ton/sandbox/dist/ui/protocol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeTransactions = serializeTransactions;
exports.serializeContracts = serializeContracts;
const core_1 = require("@ton/core");
function serializeTransactions(transactions) {
    return {
        transactions: transactions.map((t) => {
            const tx = (0, core_1.beginCell)()
                .store((0, core_1.storeTransaction)(t))
                .endCell()
                .toBoc()
                .toString('hex');
            return {
                transaction: tx,
                blockchainLogs: t.blockchainLogs,
                vmLogs: t.vmLogs,
                debugLogs: t.debugLogs,
                code: undefined,
                sourceMap: undefined,
                contractName: undefined,
                parentId: t.parent?.lt.toString(),
                childrenIds: t.children?.map((c) => c?.lt?.toString()),
                oldStorage: t.oldStorage?.toBoc().toString('hex'),
                newStorage: t.newStorage?.toBoc().toString('hex'),
                callStack: t.callStack,
            };
        }),
    };
}
function serializeContracts(contracts) {
    return contracts.map(({ contract, meta }) => {
        const state = contract.accountState;
        const stateInit = (0, core_1.beginCell)();
        if (state?.type === 'active') {
            stateInit.store((0, core_1.storeStateInit)(state.state));
        }
        const stateInitCell = stateInit.asCell();
        const account = contract.account;
        const accountCell = (0, core_1.beginCell)().store((0, core_1.storeShardAccount)(account)).endCell();
        return {
            address: contract.address.toString(),
            meta,
            stateInit: stateInitCell.bits.length === 0 ? undefined : stateInitCell.toBoc().toString('hex'),
            account: accountCell.toBoc().toString('hex'),
        };
    });
}
