// @ton/sandbox/dist/ui/UIManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UIManager = void 0;
const protocol_1 = require("./protocol");
class UIManager {
    connector;
    blockchain;
    constructor(connector, blockchain) {
        this.connector = connector;
        this.blockchain = blockchain;
    }
    publishTransactions(txs) {
        let testInfo = this.getTestInfo();
        const transactions = (0, protocol_1.serializeTransactions)(txs);
        const knownContracts = this.blockchain.knownContracts();
        const contracts = (0, protocol_1.serializeContracts)(knownContracts.map((contract) => ({ contract, meta: this.blockchain.getMeta(contract.address) })));
        this.connector.send(JSON.stringify({ type: 'test-data', testInfo, transactions, contracts }));
    }
    getTestInfo() {
        if (expect === undefined) {
            return;
        }
        const expectState = expect.getState();
        return {
            id: expectState.testId,
            name: expectState.currentTestName,
            path: expectState.testPath,
        };
    }
}
exports.UIManager = UIManager;
