// @ton/blueprint/dist/network/send/DeeplinkProvider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeeplinkProvider = void 0;
const core_1 = require("@ton/core");
const qrcode_terminal_1 = __importDefault(require("qrcode-terminal"));
const utils_1 = require("../../utils");
class DeeplinkProvider {
    constructor(network, ui) {
        this.network = network;
        this.ui = ui;
    }
    async connect() {
        return;
    }
    async sendTransaction(address, amount, payload, stateInit) {
        const deepLink = (0, utils_1.tonDeepLink)(address, amount, payload, stateInit ? (0, core_1.beginCell)().storeWritable((0, core_1.storeStateInit)(stateInit)).endCell() : undefined, this.network === 'testnet');
        try {
            this.ui.write('\n');
            qrcode_terminal_1.default.generate(deepLink, { small: true }, (qr) => this.ui.write(qr));
            this.ui.write('\n');
            this.ui.write(deepLink);
            this.ui.write('\nScan the QR code above, or open the ton:// link to send this transaction');
            await this.ui.prompt('Press enter when transaction was issued');
        }
        catch (err) {
            this.ui.write(deepLink);
            this.ui.write('\n');
            if (err instanceof Error && err.message.includes('code length overflow')) {
                this.ui.write('Message is too large to be sent via QR code. Please use the ton:// link or another method.');
                process.exit(1);
            }
            throw err;
        }
    }
    address() {
        return undefined;
    }
}
exports.DeeplinkProvider = DeeplinkProvider;
