// @ton/blueprint/dist/network/send/TonConnectProvider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TonConnectProvider = void 0;
const qrcode_terminal_1 = __importDefault(require("qrcode-terminal"));
const sdk_1 = __importDefault(require("@tonconnect/sdk"));
const core_1 = require("@ton/core");
class TonConnectStorage {
    constructor(inner) {
        this.inner = inner;
    }
    async setItem(key, value) {
        return await this.inner.setItem(key, value);
    }
    async getItem(key) {
        return await this.inner.getItem(key);
    }
    async removeItem(key) {
        return await this.inner.removeItem(key);
    }
}
function isRemote(walletInfo) {
    return 'universalLink' in walletInfo && 'bridgeUrl' in walletInfo;
}
class TonConnectProvider {
    constructor(storage, ui, network, manifestUrl = 'https://raw.githubusercontent.com/ton-org/blueprint/main/tonconnect/manifest.json') {
        this.connector = new sdk_1.default({
            storage: new TonConnectStorage(storage),
            manifestUrl,
        });
        this.ui = ui;
        this.network = network;
    }
    async connect() {
        await this.connectWallet();
        const formattedAddress = core_1.Address.parse(this.connector.wallet.account.address).toString({
            testOnly: this.network === 'testnet',
            bounceable: false,
        });
        this.ui.write(`Connected to wallet at address: ${formattedAddress}\n`);
    }
    address() {
        if (!this.connector.wallet)
            return undefined;
        return core_1.Address.parse(this.connector.wallet.account.address);
    }
    async connectWallet() {
        const wallets = (await this.connector.getWallets()).filter(isRemote);
        await this.connector.restoreConnection();
        if (this.connector.wallet) {
            return;
        }
        const wallet = await this.ui.choose('Choose your wallet', wallets, (w) => w.name);
        this.ui.setActionPrompt('Connecting to wallet...');
        const url = this.connector.connect({
            universalLink: wallet.universalLink,
            bridgeUrl: wallet.bridgeUrl,
        });
        this.ui.write('\n');
        qrcode_terminal_1.default.generate(url, { small: true }, (qr) => this.ui.write(qr));
        this.ui.write('\n' + url + '\n\n');
        this.ui.setActionPrompt('Scan the QR code in your wallet or open the link...');
        return new Promise((resolve, reject) => {
            this.connector.onStatusChange((w) => {
                if (w) {
                    resolve();
                }
                else {
                    reject('Wallet is not connected');
                }
            }, reject);
        });
    }
    async sendTransaction(address, amount, payload, stateInit) {
        this.ui.setActionPrompt('Sending transaction. Approve in your wallet...');
        const result = await this.connector.sendTransaction({
            validUntil: Date.now() + 5 * 60 * 1000,
            messages: [
                {
                    address: address.toString(),
                    amount: amount.toString(),
                    payload: payload?.toBoc().toString('base64'),
                    stateInit: stateInit
                        ? (0, core_1.beginCell)().storeWritable((0, core_1.storeStateInit)(stateInit)).endCell().toBoc().toString('base64')
                        : undefined,
                },
            ],
        });
        this.ui.clearActionPrompt();
        this.ui.write('Sent transaction');
        return result;
    }
}
exports.TonConnectProvider = TonConnectProvider;
