// @ton/blueprint/dist/network/storage/FSStorage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FSStorage = void 0;
const path_1 = __importDefault(require("path"));
const promises_1 = __importDefault(require("fs/promises"));
class FSStorage {
    constructor(path) {
        this.path = path;
    }
    async readObject() {
        try {
            return JSON.parse((await promises_1.default.readFile(this.path)).toString('utf-8'));
        }
        catch (_) {
            return {};
        }
    }
    async writeObject(obj) {
        await promises_1.default.mkdir(path_1.default.dirname(this.path), { recursive: true });
        await promises_1.default.writeFile(this.path, JSON.stringify(obj));
    }
    async setItem(key, value) {
        const obj = await this.readObject();
        obj[key] = value;
        await this.writeObject(obj);
    }
    async getItem(key) {
        const obj = await this.readObject();
        return obj[key] ?? null;
    }
    async removeItem(key) {
        const obj = await this.readObject();
        delete obj[key];
        await this.writeObject(obj);
    }
}
exports.FSStorage = FSStorage;
