// ton-assembly/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.trace = exports.text = exports.runtime = exports.logs = exports.Builder = exports.Slice = exports.Cell = void 0;
var core_1 = require("@ton/core");
Object.defineProperty(exports, "Cell", { enumerable: true, get: function () { return core_1.Cell; } });
Object.defineProperty(exports, "Slice", { enumerable: true, get: function () { return core_1.Slice; } });
Object.defineProperty(exports, "Builder", { enumerable: true, get: function () { return core_1.Builder; } });
exports.logs = __importStar(require("./logs"));
exports.runtime = __importStar(require("./runtime"));
exports.text = __importStar(require("./text"));
exports.trace = __importStar(require("./trace"));
