// @ton/blueprint/dist/ui/InquirerUIProvider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InquirerUIProvider = void 0;
const inquirer_1 = __importDefault(require("inquirer"));
const core_1 = require("@ton/core");
class DestroyableBottomBar extends inquirer_1.default.ui.BottomBar {
    destroy() {
        this.close();
    }
}
class InquirerUIProvider {
    constructor() {
        this.bottomBar = new DestroyableBottomBar();
    }
    write(message) {
        this.bottomBar.log.write(message);
    }
    async prompt(message) {
        const { prompt } = await inquirer_1.default.prompt({
            type: 'confirm',
            name: 'prompt',
            message,
        });
        return prompt;
    }
    async inputAddress(message, fallback) {
        const prompt = message + (fallback === undefined ? '' : ` (default: ${fallback})`);
        while (true) {
            const addr = (await this.input(prompt)).trim();
            try {
                return addr === '' && fallback !== undefined ? fallback : core_1.Address.parse(addr);
            }
            catch (_) {
                this.write(addr + ' is not valid!\n');
            }
        }
    }
    async input(message) {
        const { val } = await inquirer_1.default.prompt({
            name: 'val',
            message: message,
        });
        return val;
    }
    async choose(message, choices, display) {
        const { choice } = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'choice',
                message: message,
                choices: choices.map((c) => ({ name: display(c), value: c })),
            },
        ]);
        return choice;
    }
    setActionPrompt(message) {
        this.bottomBar.updateBottomBar(message);
    }
    clearActionPrompt() {
        this.bottomBar.updateBottomBar('');
    }
    close() {
        this.bottomBar.destroy();
    }
}
exports.InquirerUIProvider = InquirerUIProvider;
